com.ca.lsp.cobol/lsp-core-cobol-parser/src/test/java/com/ca/lsp/core/cobol/visitor/VisitorOutlineAnalysisTest.java

```diff
@@ -0,0 +1,175 @@
+                node("INPUT-OUTPUT SECTION", SECTION, nested(
+                    node("TRANS-FILE-IN", FILE)
+                ))
+            )),
+            node("DATA DIVISION", DIVISION, nested(
+                node("FILE SECTION", SECTION, nested(
+                    node("TRANS-FILE-IN", FILE, nested(
+                        node("FILE-RECORD", FIELD)
+                    )),
+                    node("TERMS-FILE", FILE, nested(
+                        node("TERMS-RECORD", STRUCT, nested(
+                            node("TERMS-KEY", FIELD),
+                            node("FILLER", FIELD)
+                        ))
+                    ))
+                )),
+                node("WORKING-STORAGE SECTION", SECTION, nested(
+                    node("COLR-DISPLAY", FIELD),
+                    node("CTLFILE-REC", STRUCT, nested(
+                        node("CTLFILE-PRIME", FIELD),
+                        node("CTLFILE-PAST-DUE-DIFF", FIELD)
+                    )),
+                    node("CTLFILE-REC-12", REDEFINES, nested(
+                        node("CTLFILE-DB-DATE", FIELD),
+                        node("CTLFILE-ATB-DATE", FIELD),
+                        node("COST-RECORD-CODE", FIELD, nested(
+                            node("HEADER", FIELD_88),
+                            node("SUPPLR", FIELD_88),
+                            node("WREHOUSE", FIELD_88)
+                        ))
+                    ))
+                )),
+                node("LINKAGE SECTION", SECTION, nested(
+                    node("LINK-PRM0", STRUCT, nested(
+                        node("PARM-LENGTH", FIELD),
+                        node("PARM", FIELD)
+                    ))
+                ))
+            )),
+            node("PROCEDURE DIVISION", DIVISION, nested(
+                node("000-PROGRAM-DRIVER", PROCEDURE_SECTION),
+                node("100-HOUSEKEEPING", PROCEDURE_SECTION, nested(
+                    node("110-OPEN-FILES", PROCEDURE),
+                    node("112-READ-WAREHOUS-FILE", PROCEDURE),
+                    node("199-EXIT", PROCEDURE)
+                ))
+            ))
+        ))
+    );
+    check(expectedNodes, tree.getOutlineNodes(), "");
+  }
+
+  private void check(List<OutlineNode> expected, List<OutlineNode> actual, String level) {
+    Map<String, OutlineNode> expectedMap = toMap(expected);
+    Map<String, OutlineNode> actualMap = toMap(actual);
+    assertEquals("Elements on level " + level, expectedMap.keySet(), actualMap.keySet());
+    for (String nodeName: expectedMap.keySet()) {
+      OutlineNode expectedNode = expectedMap.get(nodeName);
+      OutlineNode actualNode = actualMap.get(nodeName);
+      String newLevel = level + '/' + nodeName;
+      assertEquals("Node types for " + newLevel, expectedNode.getNodeType(), actualNode.getNodeType());
+      check(expectedNode.getNestedNodes(), actualNode.getNestedNodes(), newLevel);
+    }
+  }
```

Comment:
If I understood correctly looks like you used iteration and recursion in the same method.
I would prefer to separate the complexity for create the structure from the visiting and assert

Reply:
I've split it into two methods for a list of nodes check and for each node check.